pkg/cmd/roachtest/tests/mvcc_gc.go

```diff
@@ -0,0 +1,334 @@
+	"github.com/cockroachdb/cockroach/pkg/roachprod/install"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
+	"github.com/cockroachdb/errors"
+)
+
+func registerMVCCGC(r registry.Registry) {
+	r.Add(registry.TestSpec{
+		Name:    "mvcc_gc",
+		Owner:   registry.OwnerKV,
+		Timeout: 15 * time.Minute,
+		Cluster: r.MakeClusterSpec(3),
+		Run:     runMVCCGC,
+	})
+}
+
+/*
+Test verifies that GC respects range tombstones and garbage collects range
+tombstones and that mvcc stats are correctly updated.
+
+Test sequence:
+- create kv test data schema
+- pre-split with 3 min expiration to allow merging to proceed in parallel after
+  initial data is loaded
+- load test data using KV workload
+- delete all data in kv table using range tombstones
+  (deletion is done using cli commands that sends kv requests to cluster)
+- sets low GC ttl to make all data eligible for deletion after 2 min
+- write more data over the existing one
+- verify that all garbage and range keys were removed by GC within 5 min
+*/
+
+func runMVCCGC(ctx context.Context, t test.Test, c cluster.Cluster) {
+	c.Put(ctx, t.Cockroach(), "./cockroach")
+	s := install.MakeClusterSettings()
+	s.Env = append(s.Env, "COCKROACH_SCAN_INTERVAL=30s")
+	c.Start(ctx, t.L(), option.DefaultStartOpts(), s)
+
+	conn := c.Conn(ctx, t.L(), 1)
+	defer conn.Close()
+
+	execSQLOrFail := func(statement string, args ...interface{}) {
+		if _, err := conn.ExecContext(ctx, statement, args...); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	setClusterSetting := func(name string, value interface{}) {
+		execSQLOrFail(fmt.Sprintf(`SET CLUSTER SETTING %s = $1`, name), value)
+	}
+
+	setClusterSetting("storage.mvcc.range_tombstones.enabled", true)
+	setClusterSetting("kv.protectedts.poll_interval", "5s")
+	setClusterSetting("kv.mvcc_gc.queue_interval", "0s")
+
+	m := c.NewMonitor(ctx)
+	m.Go(func(ctx context.Context) error {
+		c.Run(ctx, c.Node(1), "./cockroach", "workload", "init", "kv", "--cycle-length", "10000")
+
+		splitKvIntoRanges(t, conn, 100, 180*time.Second, "kv", "kv")
+
+		c.Run(ctx, c.Node(1), "./cockroach", "workload", "run", "kv", "--cycle-length", "10000",
+			"--max-block-bytes", "2048", "--min-block-bytes", "2048", "--duration", "2m",
+			"--read-percent", "0")
+
+		t.L().Printf("Finished prepping GC data")
+
+		if err := retry.WithMaxAttempts(ctx, retry.Options{}, 3, func() error {
+			return deleteAllTableDataWithTombstones(ctx, t, c, conn, "kv", "kv")
+		}); err != nil {
+			t.Fatal(err)
+		}
+
+		t.L().Printf("Deleted all data using tombstones")
+
+		if err := assertTableHasNoLiveData(t, conn, "kv", "kv"); err != nil {
+			t.Fatal(err)
+		}
+
+		execSQLOrFail("alter database kv configure zone using gc.ttlseconds = $1", 120)
+
+		t.L().Printf("Configured GC ttl to 2 minutes")
+
+		c.Run(ctx, c.Node(1), "./cockroach", "workload", "run", "kv", "--duration", "2m", "--read-percent", "0")
```

Comment:
Should we add some reads here too, just to make sure they return the expected results?

Reply:
KV reads don't really validate anything, they just run blind query for key that is known already exist. It only makes sense if we change workload to derive value from key and then verify against previously written key. Currently workloads don't fail on their queries I believe unless it's a technical failure.